src/Symfony/Component/Serializer/Normalizer/ChainNormalizer.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\Serializer\Exception\LogicException;
+use Symfony\Component\Serializer\Exception\NotNormalizableValueException;
+use Symfony\Component\Serializer\Serializer;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * This holds a collection of normalizers. It tries to be smart how it selects
+ * the correct Normalizer.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class ChainNormalizer implements NormalizerInterface, SerializerAwareInterface
+{
+    /**
+     * @deprecated since Symfony 7.2
+     */
+    private ?SerializerInterface $serializer = null;
+
+    /**
+     * @var NormalizerInterface[]
+     */
+    private array $normalizers = [];
+
+    /**
+     * @var array<string, array<string, array<bool>>>
+     */
+    private array $normalizerCache = [];
+
+    /**
+     * @var array<string, array<class-string|'*'|'object'|string, bool|null>>
+     */
+    private array $supportedCache = [];
+
+    /**
+     * @param NormalizerInterface[] $normalizers
+     */
+    public function __construct(array $normalizers = [])
+    {
+        foreach ($normalizers as $normalizer) {
+            $this->addNormalizer($normalizer);
+        }
+    }
+
+    /**
+     * Add Normalizer last in the line.
+     */
+    public function addNormalizer(NormalizerInterface $normalizer): void
+    {
+        if ($normalizer instanceof NormalizerAwareInterface) {
+            $normalizer->setNormalizer($this);
```

Comment:
if we make the ChainNormalizer available publicly, this is flawed as it breaks when nesting a ChainNormalizer inside another ChainNormalizer (the normalizers inside the inner ChainNormalizer won't get access to the main normalizer supporting everything)

Reply:
Would a solution be to have `ChainNormalizer` also implement the `NormalizerAwareInterface`?